lib/mongo/socket.rb

```diff
@@ -265,18 +263,86 @@ def connectable?
 
     private
 
-    def read_from_socket(length, timeout: nil)
+    # Reads the +length+ bytes from the socket, the read operation duration is
+    # limited to +timeout+ second.
+    #
+    # @param [ Integer ] length The number of bytes to read.
+    # @param [ Numeric ] timeout The total timeout to the whole read operation.
+    #
+    # @return [ Object ] The data from the socket.
+    def read_with_timeout(length, timeout)
+      deadline = Utils.monotonic_time + timeout
+      map_exceptions do
+        socket_timeout = deadline - Utils.monotonic_time
+        if socket_timeout <= 0
+          raise Mongo::Error::TimeoutError
+        end
+        data = read_from_socket(length, socket_timeout: socket_timeout, csot: true)
+        unless (data.length > 0 || length == 0)
+          raise IOError, "Expected to read > 0 bytes but read 0 bytes"
+        end
+        while data.length < length
+          socket_timeout = deadline - Utils.monotonic_time
+          if socket_timeout <= 0
+            raise Mongo::Error::TimeoutError
+          end
+          chunk = read_from_socket(length - data.length, socket_timeout: socket_timeout, csot: true)
+          unless (chunk.length > 0 || length == 0)
+            raise IOError, "Expected to read > 0 bytes but read 0 bytes"
+          end
+          data << chunk
+        end
+        data
+      end
+    end
+
+    # Reads the +length+ bytes from the socket. The read operation may involve
+    # multiple socket reads, each read is limited to +timeout+ second,
+    # if the parameter is provided.
+    #
+    # @param [ Integer ] length The number of bytes to read.
+    # @param [ Numeric ] socket_timeout The timeout to use for each chunk read.
+    #
+    # @return [ Object ] The data from the socket.
+    def read_without_timeout(length, socket_timeout = nil)
+      map_exceptions do
+        data = read_from_socket(length, socket_timeout: socket_timeout)
+        unless (data.length > 0 || length == 0)
+          raise IOError, "Expected to read > 0 bytes but read 0 bytes"
+        end
+        while data.length < length
+          chunk = read_from_socket(length - data.length, socket_timeout: socket_timeout)
+          unless (chunk.length > 0 || length == 0)
+            raise IOError, "Expected to read > 0 bytes but read 0 bytes"
+          end
+          data << chunk
+        end
+        data
```

Comment:
I know this was just copied from the original implementation of `#read`, but we can probably get rid of the duplication as was suggested above for `read_with_timeout`, by letting the loop do all the work:

```suggestion
        "".tap do |data|
          while data.length < length
            chunk = read_from_socket(length - data.length, socket_timeout: socket_timeout)
            unless chunk.length > 0
              raise IOError, "Expected to read > 0 bytes but read 0 bytes"
            end
            data << chunk
          end
        end
```